### DISCUSSION TITLE
Nuovo ordine - Il totale dei pagamenti non corrisponde al totale da pagare su ordine senza totali

### DISCUSSION THREAD

**Discussion by stefanocapitanio** [fattureincloud/api#467] [Q&A] [answered]
Ciao, spero mi possiate aiutare. Ho fatto un po' di ricerche, ho trovato una segnalazione per quanto riguarda le fatture. Però non ho trovato riferimenti per gli ordini.
Di seguito la nostra request di creazione di un ordine che va in errore:
```
{
  "companyId": 111111,
  "createIssuedDocumentRequest": {
    "data": {
      "entity": {
        "id": 111111,
        "name": "---",
        "tax_code": "---",
        "address_province": "MI",
        "address_postal_code": "20061",
        "address_city": "Carugate",
        "address_street": "---"
      },
      "type": "order",
      "date": "2024-08-19",
      "visible_subject": "test",
      "notes": "test",
      "subject": "Riferimento intervento xxx.it",
      "items_list": [
        {
          "product_id": "34177606",
          "code": "RL-INT",
          "name": "Ore di Intervento",
          "description": "ORE DI LAVORO 19/08/2024",
          "measure": "ore",
          "net_price": 45,
          "qty": 5
        },
        {
          "product_id": "34177611",
          "code": "RL-KM-VIAGGIO",
          "name": "Km Viaggio Intervento",
          "description": "KM DI VIAGGIO 19/08/2024",
          "measure": "km",
          "net_price": 0.5,
          "qty": 10
        },
        {
          "product_id": "34177610",
          "code": "RL-ORE-VIAGGIO",
          "name": "Ore Viaggio Intervento",
          "description": "ORE DI VIAGGIO 19/08/2024",
          "measure": "ore",
          "net_price": 30,
          "qty": 1
        },
        {
          "code": "LUXONE03000",
          "description": "INTERRUTTORE",
          "qty": 1,
          "net_price": 30,
          "measure": "n",
          "product_id": "34177540"
        }
      ]
    }
  }
}
```
La risposta che arriva è:
```
{
  "status": 422,
  "body": {
    "error": {
      "message": "Il totale dei pagamenti non corrisponde al totale da pagare.",
      "validation_result": null
    },
    "extra": {
      "totals": {
        "amount_net": 290,
        "amount_global_cassa_taxable": 290,
        "taxable_amount": 290,
        "vat_list": [
          {
            "amount_net": 290,
            "amount_vat": 0
          }
        ],
        "amount_vat": 0,
        "amount_gross": 290,
        "amount_enasarco_taxable": 0,
        "amount_due": 290,
        "amount_due_discount": null,
        "payments_sum": 0
      }
    }
  },
...
```
Non abbiamo messo la parte dei totals perché con il tipo documento Rapporto di intervento funzionava anche senza. Inoltre non stiamo gestendo l'IVA che pensavamo di demandare al gestionale.
So che con la fattura potrebbe essere un problema, però con l'ordine pensavo non lo fosse.
Come dobbiamo procedere? Che documenti prevedono il calcolo del totale e che documenti non lo prevedono?

Grazie

**Comment by emajo** [comment]
Ciao @stefanocapitanio i tipi di documento `ordini`, `proforma`, `fatture`, `autofatture`, `ricevute` e `note di credito` richiedono l'inserimento della sezione relativa ai pagamenti (che possono essere saldati o meno), mentre `ordini a fornitore`, `rapporti d'intervento`, `preventivi` e `ddt` non richiedono i pagamenti.

Come procedere dipende molto da quale è l'obbiettivo del documento che sta cercando di creare, banalmente basterebbe mettere un pagamento non saldato se questo è il caso.

Grazie,
Emanuele.

**Reply by stefanocapitanio** [reply]
Grazie mille per la pronta risposta.
Per i tipi documento che funzionano senza pagamenti, OK.

L'obiettivo sarebbe: caricare il documento che preferisce l'utente su fatture in cloud con più dati possibili e poi farlo gestire da di lì.
I documenti che gestiamo attualmente sono:
 - ✅ DDT
 - ✅ Rapporto di intervento
 - ❌ Ordini
 - ❌ Fatture

Volendo gestire anche quelli, come devo procedere? Ho guardato meglio in documentazione e ho visto:

> If you don't mind about the amounts of each payment, but you just want them to match the expected totals, we provide a magic field that could be perfect for you. The **options.fix_payments** field, when set to true, adjusts your last payment amount of the payments list to match your document total, incrementing or reducing the amount accordingly.
https://developers.fattureincloud.it/docs/guides/invoice-totals/#tophat-the-magic-field

Se impostassi questo flag, dovrei comunque mandare in request anche la payment list?
```
"payments_list": [
      {
        "amount": null,
        "due_date": null,
        "paid_date": null,
        "status": "not_paid"
      }
    ]
```

Facendo così il documento verrebbe generato correttamente?

Grazie

**Reply by emajo** [reply, marked as answer]
Ciao @stefanocapitanio, utilizzando quel flag il campo `payments_list` può essere omesso completamente, ci penserà la nostra api a generarlo in relazione ai totali degli articoli.

Grazie,
Emanuele.

**Reply by stefanocapitanio** [reply]
Fantastico, grazie!
